### ISSUE TITLE
MappedColumn fails: "Expected an option type, found Mapped[Option[String']"

### ISSUE THREAD

**Issue by helllamer** [slick/slick#2246]
### === [TEST-CASE COMMIT HERE](https://github.com/helllamer/slick/commit/ba8193edb783164b9f2e319f4b238e3c3ee22431) ===

**DESCRIPTION:**
Using of mapped column in query causes query compiler to crash on slick v3.3.3.

Test-case source code part:
```
//...
class TheItemsTable(tag: Tag) extends Table[TheItem](tag, "MappedTest0") {
  //...
  def geoShapeStringOption = column[Option[String]]( "geo_shape" )
  def geoShapeOption = geoShapeStringOption <> (IGeoShape.applyOption, IGeoShape.unapplyOption)
}

theItemsQuery
  .filter(_.geoShapeStringOption.nonEmpty )
  .map(_.geoShapeOption)                                // <<< It crashes here
  .distinct
  .result
```

**STEPS TO REPRODUCE:**

```
$ git fetch https://github.com/helllamer/slick
$ git checkout ba8193edb783164b9f2e319f4b238e3c3ee22431
$ sbt
> test
```

**EXPECTED RESULT:**
Testkit must success.

**ACTUAL RESULT:**
First test in test suite fails with:
```
[info] Test com.typesafe.slick.testkit.tests.MappedColumnTest.test[sqlitedisk] started
[error] Test com.typesafe.slick.testkit.tests.MappedColumnTest.test[sqlitedisk] failed:
 slick.SlickException: Expected an option type, found Mapped[Option[String']], took 0.175 sec
[error]     at slick.ast.TypeUtil$.asOptionType$extension(Type.scala:288)
[error]     at slick.ast.OptionFold.withInferredType(Node.scala:692)
[error]     at slick.ast.OptionFold.withInferredType(Node.scala:682)
[error]     at slick.ast.Node.infer(Node.scala:89)
[error]     at slick.ast.Node.infer$(Node.scala:88)
[error]     at slick.ast.OptionFold.infer(Node.scala:682)
[error]     at slick.ast.SimplyTypedNode.$anonfun$withInferredType$1(Node.scala:122)
[error]     at slick.ast.UnaryNode.mapChildren(Node.scala:223)
[error]     at slick.ast.UnaryNode.mapChildren$(Node.scala:221)
[error]     at slick.ast.TypeMapping.mapChildren(Node.scala:715)
[error]     at slick.ast.SimplyTypedNode.withInferredType(Node.scala:122)
[error]     at slick.ast.SimplyTypedNode.withInferredType$(Node.scala:121)
[error]     at slick.ast.TypeMapping.withInferredType(Node.scala:715)
[error]     at slick.ast.TypeMapping.withInferredType(Node.scala:715)
[error]     at slick.ast.Node.infer(Node.scala:89)
[error]     at slick.ast.Node.infer$(Node.scala:88)
[error]     at slick.ast.TypeMapping.infer(Node.scala:715)
[error]     at slick.ast.ComplexFilteredQuery.$anonfun$withInferredType$2(Node.scala:303)
[error]     at slick.ast.BinaryNode.mapChildren(Node.scala:204)
[error]     at slick.ast.BinaryNode.mapChildren$(Node.scala:200)
[error]     at slick.ast.Distinct.mapChildren(Node.scala:406)
[error]     at slick.ast.ComplexFilteredQuery.withInferredType(Node.scala:302)
[error]     at slick.ast.ComplexFilteredQuery.withInferredType(Node.scala:295)
[error]     at slick.ast.Node.infer(Node.scala:89)
[error]     at slick.ast.Node.infer$(Node.scala:88)
[error]     at slick.ast.FilteredQuery.infer(Node.scala:283)
[error]     at slick.compiler.ExpandTables.$anonfun$apply$3(ExpandTables.scala:73)
[error]     at slick.ast.ClientSideOp$.mapServerSide(ClientSideOp.scala:20)
[error]     at slick.compiler.ExpandTables.$anonfun$apply$2(ExpandTables.scala:36)
[error]     at slick.compiler.CompilerState.map(QueryCompiler.scala:230)
[error]     at slick.compiler.ExpandTables.apply(ExpandTables.scala:36)
[error]     at slick.compiler.QueryCompiler.$anonfun$runPhase$1(QueryCompiler.scala:79)
[error]     at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
[error]     at slick.ast.SymbolNamer.use(Symbol.scala:111)
[error]     at slick.compiler.QueryCompiler.runPhase(QueryCompiler.scala:78)
[error]     at slick.compiler.QueryCompiler.$anonfun$runPhases$9(QueryCompiler.scala:75)
[error]     at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:646)
[error]     at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:642)
[error]     at scala.collection.immutable.NewVectorIterator.foldLeft(Vector.scala:1966)
[error]     at slick.compiler.QueryCompiler.runPhases(QueryCompiler.scala:75)
[error]     at slick.compiler.QueryCompiler.run(QueryCompiler.scala:53)
[error]     at slick.compiler.QueryCompiler.run(QueryCompiler.scala:47)
[error]     at slick.basic.BasicProfile$API.streamableQueryActionExtensionMethods(BasicProfile.scala:56)
[error]     at slick.basic.BasicProfile$API.streamableQueryActionExtensionMethods$(BasicProfile.scala:55)
[error]     at slick.jdbc.JdbcProfile$$anon$1.streamableQueryActionExtensionMethods(JdbcProfile.scala:67)
[error]     at com.typesafe.slick.testkit.tests.MappedColumnTest.$anonfun$test$1(MappedColumnTest.scala:52)
[error]     at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:672)
[error]     at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
[error]     at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1429)
[error]     at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[error]     at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
[error]     at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
[error]     at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
[error]     at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**SETUP:**
No special setup.
Databases/profiles/configuration does not matter.
Problem related to query compilation.

**WORKAROUND:**
Rollback to `slick-3.3.2`.
On 3.3.2 and several years before was no issue here.

**Comment by nafg** [comment]
Hi, what would be really amazing if if you could find the exact commit that
broke it, perhaps with git bisect

**Comment by helllamer** [comment]
@nafg `git bisect` pointed to [this commit](https://github.com/slick/slick/commit/f81a69c798b76bb44e16045abddbd9c8922f5f64) by @shmishleniy .

**Comment by nafg** [comment]
@shmishleniy any chance you can look into this?

**Comment by shmishleniy** [comment]
you can use .distinctOn as a workaround

```
theItemsQuery
  .filter(_.geoShapeStringOption.nonEmpty )
  .distinctOn(_.geoShapeStringOption)
  .map(_.geoShapeOption)                            
  .result
```

**Comment by nafg** [comment]
Closed by #2248
